intro-to-oop-contd/testing-objects.md

```diff
@@ -0,0 +1,263 @@
+| --- | --- | --- | ---
+
+## Unit Tests Verify State and Behavior
+
+Recall that classes and objects are responsible for tracking state and behavior. Therefore, our unit tests on class should be focused on:
+
+- For all different situations, are the attributes of an instance correct?
+- For all different situations, do the methods on an instance work as expected?
+
+We can expand on these questions and come up with many generic test cases:
+
+- When we create an instance of this class, are the initial attributes set to what's expected?
+    - Can we read the attributes and get back the expected value?
+- Can an instance of this class call an instance method, and observe expected behavior?
+    - Can we call an instance method with specific parameters?
+        - Do we get back the return value we expect?
+        - Do any of the instance's attributes change?
+        - Do any of the instance's attributes explicitly stay the same?
+
+As always, the actual test cases depend on the context of the class and feature. Some classes don't have much on state or behavior. Some classes have very complex logic which need to be thoroughly tested.
+
+### Example
+
+Scarlet is making a ride share app, which has three classes, `Driver`, `Passenger`, and `Trip`. Every driver has a name, VIN, and list of trips they've driven. Every driver can calculate their average rating and add a new trip.
+
+Scarlet will likely test the `Driver` class's initial state by checking:
+
+- Initial value `name`
+- Initial value of `vin`
+- Initial value of `trips`
+
+She can test an instance method named `calculate_avg_rating` by checking:
+
+- Return value of `calculate_avg_rating` for a `Driver` instance with 2+ trips
+- Return value of `calculate_avg_rating` for a `Driver` instance with 0 trips
+
+She can test an instance method named `add_trip` by checking:
+
+- When passing a valid `new_trip` into the method...
+    - Return value of `add_trip`
+    - The state of the `trips` attribute should change
+        - The length is increased
+        - The `trips` list contains the `new_trip`
+- When passing an invalid `new_trip` into the method...
+    - Return value of `add_trip`
+    - The state of the `trips` attribute should **not** change
+        - The length is the same
+        - The `trips` list does not contain the `new_trip`
+
+### Exercise: Test Cases
+
+Read through this project description. Then, make a list of all relevant test cases. This list should mirror Scarlet's list of tests for her ride share app.
+
+Imagine an app that manages the customers and orders of a grocery store. There are two classes, `Customer` and `Order`.
+
+Each `Customer` has:
+    - `name`
+    - `email_address`
+
+Each `Order` has:
+    - `products`
+    - `customer`
+    - `add_product` method
+    - `remove_product` method
+
+Imagine and make up assumptions about the logic of `add_product` and `remove_product`, particularly around the **edge cases**.
```

Comment:
I'm being a weirdo: Gonna fight for "Imagine and make up" for the first instance, and then conform to "Make up reasonable" for the rest

Reasoning:
Similar to what I've mentioned about the word "decide," "make up assumptions/make up reasonable assumptions" somehow comes across as closer to "figure out the assumptions," as if saying "there is a correct answer to this that you should be able to conclude within this exercise."

In my opinion, adding the prefix "Imagine and" softens it and opens it up to being like, "hey, if imagination is at play, there isn't a correct answer. We're in the middle of an extended metaphor, and playin' in the exploration space."

(to be clear, a lot of this is because contextually we're introducing new concepts, and this task is particularly creative around the new concept. If this were something like "whiteboard interview practice," I'd be comfortable saying "make up reasonable assumptions")

Reply:
If you'd like to use imagine (which is definitely softer), how about 'Imagine some possible assumptions'?

For me, 'imagine' has a vagueness to it that demands a bit more lead into the thing being imagined.  For example, consider distributing the object (assumptions) over the verbs, and we get "imagine assumptions" and "make up assumptions".  Both sound blunt to me.  I feel like the make up version is a little simpler to adjust with a little adjective (I chose reasonable, but possible works just as well—anything to put a little breathing room between the verb and the object here).

I also associate imagine as more like "picture in your mind a possible representation of a thing." I more let the concept come to me rather than actively driving its representation.  Whereas "make up" involves guiding the experience more actively.  So imagine picturing a possibly existing case, while making up is more of an act of creation.  Does that make any sense???  I dunno.

So really for imagine, I feel like I want even a little _more_ room, since imagination is a freeform concept. So to expand and distinguish the subtle difference between imagining and making up, maybe something like:

```suggestion
Imagine how the `add_product` and `remove_product` methods might behave, and make up some possible  assumptions about their logic, particularly around the **edge cases**.
```